Migrate OC_Mail to SwiftMailer

LukasReschke opened the pull request:
This refactors OC_Mail to use SwiftMailer instead of PHPMail, furthermore this fixes the unit tests when running locally due to being able to mock the  mailing class. (#12014)

Requires https://github.com/owncloud/3rdparty/pull/143

@PVince81 @th3fallen @bantu @DeepDiver1975 Please review carefully. - Thanks!

Todos
======
- [x] Revert #12489

nickvergessen commented:
Another problem with PHPMailer:
https://github.com/owncloud/core/issues/13252

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cdd9df25-2de4-4fa2-8820-b372d0e7c681) was created.

LukasReschke commented:
Would love to get some momentum on this. I rebased it and added some more unit tests for code that wasn't unit-testable before. We need this as well for one of our blue things ;-)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4efd5a47-6015-4c48-9dbd-290d6664bb71) was created.

LukasReschke commented:
Usage is now like: 

```php
$message = $this->mailer->createMessage();
$message->setTo([$email => $username]);
$message->setSubject($subject);
$message->setHtmlBody($mailContent);
$message->setPlainBody($plainTextMailContent);
$message->setFrom([$this->fromMailAddress => $this->defaults->getName()]);
$this->mailer->send($message);
```

LukasReschke commented:
This is also crucial to be able to test our code… Currently it's impossible to write unit tests for anything that uses OC_Mail… :(

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/cdd9df25-2de4-4fa2-8820-b372d0e7c681) completed: **26 new issues, 66 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/4efd5a47-6015-4c48-9dbd-290d6664bb71) completed: **26 new issues, 66 updated code elements**

th3fallen commented on core/lostpassword/controller/lostcontroller.php:
```diff
@@ -201,7 +201,7 @@ protected function sendEmail($user) {
 		try {
 			// FIXME: should be added to the container and injected in here
-			\OC_Mail::send(
+			\OCP\Util::sendMail(
```
Erm... This is depreciated and not using your new class...

LukasReschke replied:
Fair point. Will adjust as well and add unit tests then as well.

LukasReschke replied:
Done with https://github.com/owncloud/core/commit/bff946cbdf38feafac30c46543ba828d720e1262

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c31dd76c-3f2d-4bc3-9a65-02fe110139a3) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c31dd76c-3f2d-4bc3-9a65-02fe110139a3) completed: **21 new issues, 68 updated code elements**

MorrisJobke commented:
@LukasReschke Have a look at scrutinizer ;)

LukasReschke commented:
@MorrisJobke Why?

https://scrutinizer-ci.com/g/owncloud/core/inspections/c31dd76c-3f2d-4bc3-9a65-02fe110139a3/issues/?status=new&selectedAuthors%5B0%5D=lukas%40owncloud.com&orderField=path&order=asc

Those "bugs" are caused by old code which was copied 1:1 I don't see any reason to check if a template actually succeeded or not. If it fails it fails and it shouldn't (especially since the unit tests are testing this anyways)

LukasReschke commented:
And Code Duplication in the tests… Well there is worse ;-)

DeepDiver1975 commented:
I agree to the current interface design - need to find the time to review ....

MorrisJobke commented:
https://scrutinizer-ci.com/g/owncloud/core/inspections/c31dd76c-3f2d-4bc3-9a65-02fe110139a3/issues/files/core/lostpassword/controller/lostcontroller.php?status=new&selectedAuthors%5B0%5D=lukas%40owncloud.com&orderField=path&order=asc ;)

LukasReschke commented:
> Those "bugs" are caused by old code which was copied 1:1 I don't see any reason to check if a template actually succeeded or not. If it fails it fails and it shouldn't (especially since the unit tests are testing this anyways)

MorrisJobke commented:
Okay :P

LukasReschke commented:
:innocent:

LukasReschke commented:
@DeepDiver1975 Found anytime to review? – Rebasing does not really make sense before you have found time as I don't want to rebase this for the 40th time ;-)

DeepDiver1975 commented on lib/public/mail/util.php:
```diff
@@ -0,0 +1,26 @@
+	 * @param string $email Email address to be validated
+	 * @return bool True if the mail address is valid, false otherwise
+	 */
+	public static function validateMailAddress($email) {
```
bääh - static Kung-Fu - can we keep this within the Mailer class - does that make sense?

LukasReschke replied:
Sure. Will move it there.

DeepDiver1975 commented on tests/lib/mail.php:
```diff
@@ -1,53 +0,0 @@
-			array('info@example.com', 'info@example.com'),
-			array('info@xn--cjr6vy5ejyai80u.com', 'info@國際化域名.com'),
-			array('info@xn--mller-kva.de', 'info@müller.de'),
-			array('info@xn--mller-kva.xn--mller-kva.de', 'info@müller.müller.de'),
```
hey - I want my müller-müller address back ;-)

DeepDiver1975 commented:
@LukasReschke you have my :+1: on this - did ever anybody really send an email with this pr?

:hide:

LukasReschke commented:
> did ever anybody really send an email with this pr?

Real men don't test their changes ;-)

Yes – tested it locally – though a while ago and probably 5 rebases ago. Worked firn at this time ;-)

DeepDiver1975 commented:
> Real men don't test their changes ;-)

![59363684](https://cloud.githubusercontent.com/assets/1005065/6270372/b83eff66-b859-11e4-8e44-51c2c4f9c6d4.jpg)

MorrisJobke commented:
Ping me once it is rebased and I will review and test with real mails ;)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1efa02f2-8f2c-4f9b-ad58-61253a450624) was created.

LukasReschke commented:
@MorrisJobke Done… I'm pretty sure I fucked up some 3rdparty ref so if you could update it this would be awesome *hides* ;-)

*****PING****** ;-)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/1efa02f2-8f2c-4f9b-ad58-61253a450624) completed: **24 new issues, 62 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f8e631b4-8cf5-45ef-837f-7ee7f5772653) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f8e631b4-8cf5-45ef-837f-7ee7f5772653) completed: **24 new issues, 62 updated code elements**

DeepDiver1975 commented:
> @MorrisJobke Done… I'm pretty sure I fucked up some 3rdparty ref so if you could update it this would be awesome hides ;-)
> 
> **PING*** ;-)

let me have a closer look ...

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/23cfa584-9e38-435e-920d-26dccb5a5dc2) was created.

DeepDiver1975 commented:
@LukasReschke I took the freedom to rebase the branch

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/23cfa584-9e38-435e-920d-26dccb5a5dc2) completed: **24 new issues, 62 updated code elements**

DeepDiver1975 commented:
I just tried to connect with two different smtp servers .... without success

error message somewhere in the depth:
````
Expected response code 250 but got code "", with message ""

LukasReschke commented:
What did you configure exactly?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/5cd441b2-19e5-4a0b-a00d-7e4c77df869d) was created.

LukasReschke commented:
So, in some cases `SwiftMailer` seems to have a problem connecting to the remote host when the connection hasn't been started manually.

I added this and modified the email test wizard as well to show a more technical error message in case something happens.

LukasReschke commented:
If you want to try with your GMail account you might need to allow login in https://www.google.com/settings/security/lesssecureapps?pli=1 as well

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/5cd441b2-19e5-4a0b-a00d-7e4c77df869d) completed: **26 new issues, 60 updated code elements**

LukasReschke commented:
And another round of rebase required… – @DeepDiver1975 Let me know if the current approach works then I'll rebase :see_no_evil:

MorrisJobke commented:
@LukasReschke @DeepDiver1975 Lets get this in. Feel free to ping me, we need this.

DeepDiver1975 commented:
@MorrisJobke we have to ensure that existing functional mail setting continue to work with swiftmailer - I failed to make this work the other day ......

DeepDiver1975 commented:
Testplan:
- [ ] use google smtp
- [ ] use owncloud company smtp
- [ ] use some self hosted smtp with selfsigned cert

MorrisJobke commented:
I can test the first two, but I don't have a SMTP host with a self-signed cert.

MorrisJobke commented:
@LukasReschke Rebase fun?

LukasReschke commented:
Rebased. Hopefully all went right…

What works for me:

- SMTP with Gmail
- SMTP with oc.com SMTP

Didn't test with this checkout:
- Plain PHP + sendmail

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b48d93d3-725b-4e74-b43d-507676865736) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b48d93d3-725b-4e74-b43d-507676865736) completed: **No issues found**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/b48d93d3-725b-4e74-b43d-507676865736) completed: **No issues found**

MorrisJobke commented:
* [ ] HTML/Plain mails are shown wrong. It seems that there happens something that bricks the multipart mail

MorrisJobke commented:
BTW: this worked before. Also there is now no log entry anymore, that a mail was send (debug level)

LukasReschke commented:
>  HTML/Plain mails are shown wrong. It seems that there happens something that bricks the multipart mail

Works fine here. So there is nothing left than:

![](http://www.quickmeme.com/img/f8/f8b63f208e37c2e83be5988996caf9d8002bf5c875aacc2e017a879b1640dd0d.jpg)

Used code:
```php

	$mailer = \OC::$server->getMailer();
	$message = $mailer->createMessage();
	$message->setTo(['lukas@owncloud.com']);
	$message->setPlainBody('My plain body');
	$message->setHtmlBody('<strong>My HTML Body</strong>');
	$mailer->send($message);
```

It also works fine when only sending an HTML or Plain Body…

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e92d0c5b-f291-4110-a45b-cacdba628a4e) was created.

LukasReschke commented:
> BTW: this worked before. Also there is now no log entry anymore, that a mail was send (debug level)

Log message is back.

MorrisJobke commented:
@LukasReschke Try to send the share link via mail:

Before
![bildschirmfoto von 2015-03-16 13-48-33](https://cloud.githubusercontent.com/assets/245432/6666365/e78ed744-cbe3-11e4-9108-3c130a949a69.png)
(the image is broken because it's a localhost URL)

After
![bildschirmfoto von 2015-03-16 13-48-10](https://cloud.githubusercontent.com/assets/245432/6666366/eba77606-cbe3-11e4-9e78-e5b03dcbcb2b.png)

This is the same with Thunderbird and Google Mail. So I would say an problem at our side rather than the mail client ;)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e92d0c5b-f291-4110-a45b-cacdba628a4e) completed: **27 new issues, 60 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e97ea348-11a1-4b2d-a36a-593e2b676174) was created.

LukasReschke commented:
Should be fixed. That was caused by the legacy `OCP\Util::sendMail` which had a really confusing signature and I was not sure what `$altbody` was for. (I assumed it would be the HTML text while it is the HTML one)

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/e97ea348-11a1-4b2d-a36a-593e2b676174) completed: **27 new issues, 61 updated code elements**

MorrisJobke commented:
Works now :+1:

nickvergessen commented:
:+1: tested with the activity emails and added some html to them.
